File: app/src/main/java/kr/djspi/pipe01/nfc/NfcUtil.java
package kr.djspi.pipe01.nfc;

import android.app.Activity;
import android.app.PendingIntent;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.IntentFilter.MalformedMimeTypeException;
import android.nfc.NdefMessage;
import android.nfc.NdefRecord;
import android.nfc.NfcAdapter;
import android.nfc.Tag;
import android.nfc.tech.MifareUltralight;
import android.nfc.tech.Ndef;
import android.nfc.tech.NfcA;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.util.Log;

import com.nxp.nfclib.NxpNfcLib;
import com.nxp.nfclib.exceptions.NxpNfcLibException;
import com.nxp.nfclib.ndef.NdefMessageWrapper;
import com.nxp.nfclib.ndef.NdefRecordWrapper;
import com.nxp.nfclib.ntag.INTag213215216;
import com.nxp.nfclib.ntag.NTagFactory;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.math.BigInteger;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import static android.nfc.NdefRecord.RTD_TEXT;
import static android.nfc.NdefRecord.TNF_WELL_KNOWN;
import static com.nxp.nfclib.CardType.NTag216;
import static kr.djspi.pipe01.BuildConfig.APPLICATION_ID;
import static kr.djspi.pipe01.BuildConfig.NFC_LICENSE_KEY;
import static kr.djspi.pipe01.BuildConfig.a;
import static kr.djspi.pipe01.BuildConfig.k;
import static kr.djspi.pipe01.BuildConfig.setReadOnly;

public final class NfcUtil {

    private static final String TAG = NfcUtil.class.getSimpleName();
    private static final char[] HEX_ARRAY = "0123456789ABCDEF".toCharArray();
    private static NfcAdapter nfcAdapter;
    private IntentFilter[] intentFilters;
    private NxpNfcLib nxpNfcLib;
    private static INTag213215216 objNtag;
    /**
     * 아래의 변수들은 반드시 final 선언해야만 하며, 그렇지 않을 경우 intent 들 간의 간섭이 발생하여
     * NFC 태그를 태깅하면 이전 액티비티 인텐트를 실행하기도 한다.
     */
    private final PendingIntent pendingIntent;
    private final String[][] techLists;
    private final Activity activity;

    public NfcUtil(Activity activity, Class<?> useActivityClass) {
        this.activity = activity;
        nfcAdapter = NfcAdapter.getDefaultAdapter(this.activity);

        try {
            intentFilters = new IntentFilter[]{new IntentFilter(NfcAdapter.ACTION_NDEF_DISCOVERED, "text/plain")};
            intentFilters[0].addDataType("application/kr.djspi.pipe01");
        } catch (MalformedMimeTypeException ignore) {
        }
        pendingIntent = PendingIntent.getActivity(this.activity, 0, new Intent(this.activity, useActivityClass).setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP), 0);
        techLists = new String[][]{{NfcA.class.getName()}, {Ndef.class.getName()}, {MifareUltralight.class.getName()}};

        initializeLibrary(this.activity);
    }

    /**
     * TapLinx NTAG 라이브러리를 불러옴
     */
    private void initializeLibrary(Activity activity) {
        try {
            nxpNfcLib = NxpNfcLib.getInstance();
            nxpNfcLib.registerActivity(activity, NFC_LICENSE_KEY);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage());
        }
    }

    public static Tag onNewTagIntent(@NotNull Intent intent) {
        return intent.getParcelableExtra(NfcAdapter.EXTRA_TAG);
    }

    /**
     * NTAG 의 모델명을 찾은 뒤 적절한 태그 객체를 리턴
     * (NTAG 216 이 아님) 에러 메시지 출력, null 리턴
     *
     * @param intent 발생한 NTAG 인텐트
     * @return objNtag 인식된 NTAG 216 태그 객체
     */
    @Nullable
    private INTag213215216 getTagType(final Intent intent) {
        try {
            if (nxpNfcLib.getCardType(intent) == NTag216) {
                objNtag = NTagFactory.getInstance().getNTAG216(nxpNfcLib.getCustomModules());
            } else {
                Log.w(TAG, "TAG is NOT NTAG 216 Type");
                return null;
            }
        } catch (NxpNfcLibException e) {
            e.getMessage();
        }
        return objNtag;
    }

    /**
     * 쓰기 대상 태그에 쓰기 작업 수행
     * (setReadOnly) build.gradle 에서 설정: release 버전(true)과 내부용 alpha 버전(false) 구분해 빌드
     *
     * @param intent 전달된 태그 인텐트
     * @return isSuccess 쓰기 작업 성공 여부
     * @see NdefRecordWrapper[] createRecord 사용자 입력값으로 NDEF 레코드를 생성해 리턴
     */
    public boolean writeTag(final Intent intent, String[] strings) {
        boolean isSuccess = false;
        try {
            objNtag = getTagType(intent);
            if (objNtag == null) return false;
            objNtag.getReader().connect();
            objNtag.getReader().setTimeout(2000);
            objNtag.authenticatePwd(k, a); // 비밀번호 인증
            if (objNtag.isPwdAuthenticated()) {
                // NDEF 메시지 생성 & 쓰기 작업 실행
                objNtag.writeNDEF(new NdefMessageWrapper(createRecord(Locale.KOREAN, true, strings)));
                Log.w(TAG, "NTag Written");
                if (setReadOnly) objNtag.makeCardReadOnly(); // 쓰기 후 Read-Only 로 설정
                isSuccess = true;
            }
        } catch (NullPointerException | NxpNfcLibException | IllegalArgumentException e) {
            Log.e(TAG, "Exception Thrown: writeTag()");
//            Toast.makeText(context, R.string.toast_error, Toast.LENGTH_LONG).show();
            e.getMessage();
        }
        objNtag.getReader().close();
        return isSuccess;
    }

    /**
     * (TapLinx 라이브러리) NDEF 레코드 생성함수
     * 0번 레코드(고정): SPI 도시가스 방호철판앱 패키지명("com.kr.djspi.gas01") 지정. SPI 정품 태그를 태깅할 시
     * 앱이 설치되어 있으면 앱 실행, 앱이 설치되어 있지 않으면 플레이스토어 설치페이지로 연결됨
     * 1번 레코드~: SPI 정보
     * 마지막 레코드(고정): 태깅된 NFC 칩의 UID(시리얼번호) 지정
     *
     * @param locale       텍스트 데이터 로컬 설정: (ko)
     * @param encodeInUtf8 UTF-8 인코딩으로 기록
     * @return ndefRecordWrappers 태그에 기록되는 레코드 Array
     */
    @SuppressWarnings("SameParameterValue")
    private static NdefRecordWrapper[] createRecord(Locale locale, boolean encodeInUtf8, String[] records) {
        Charset utfEncoding;
        if (encodeInUtf8) {
            utfEncoding = Charset.forName("UTF-8");
        } else {
            utfEncoding = Charset.forName("UTF-16");
        }

        final int recordLength = records.length;
        byte[][] textBytes = new byte[recordLength][];
        byte[][] data = new byte[recordLength][];
        byte[] langBytes = locale.getLanguage().getBytes(Charset.forName("UTF-8"));
        int langBytesLength = langBytes.length;
        //noinspection PointlessArithmeticExpression
        char status = (char) (0 + langBytesLength);

        NdefRecordWrapper[] ndefRecordWrappers = new NdefRecordWrapper[recordLength + 1];

        for (int i = 0; i < recordLength; i++) {
            textBytes[i] = records[i].getBytes(utfEncoding);
            data[i] = new byte[1 + langBytesLength + textBytes[i].length]; // 0번 레코드를 위해 1자리 추가
            data[i][0] = (byte) status;
            System.arraycopy(langBytes, 0, data[i], 1, langBytesLength);
            System.arraycopy(textBytes[i], 0, data[i], 1 + langBytesLength, textBytes[i].length);
            ndefRecordWrappers[i + 1] = new NdefRecordWrapper(TNF_WELL_KNOWN, RTD_TEXT, BigInteger.valueOf(i).toByteArray(), data[i]);
        }
        // 1번 레코드부터 생성하는 for 문이 끝난 뒤 0번 레코드(패키지명)를 생성한다.
        ndefRecordWrappers[0] = new NdefRecordWrapper(NdefRecord.createApplicationRecord(APPLICATION_ID));
        return ndefRecordWrappers;
    }

    /**
     * SPI 태그에 기록된 정보를 불러옴
     * (정보가 기록되지 않음) null 을 리턴
     *
     * @return 기록된 정보를 담은 ArrayList<String>
     * @throws NullPointerException (태그 없음) 정보를 가져오는 과정에서 null 참조
     */
    public static ArrayList<String> getRecord(@NotNull Intent intent) throws NullPointerException {
        Parcelable[] rawMsgs = intent.getParcelableArrayExtra(NfcAdapter.EXTRA_NDEF_MESSAGES);
        List<ParsedRecord> records = NdefMessageParser.parse((NdefMessage) rawMsgs[0]);
        ArrayList<String> recordList = new ArrayList<>();
        recordList.add(0, null);
        final int listLength = records.size() + 1;
        for (int j = 1; j < listLength; j++) {
            recordList.add(j, ((TextRecord) records.get(j - 1)).getText());
        }
        return recordList;
    }

    /**
     * 전달 받은 Tag 에서 데이터를 가져와 리턴
     *
     * @param tag
     * @return
     */
    @NotNull
    public String readTag(Tag tag) {
        StringBuilder ret = new StringBuilder();
        try {
            Ndef ndefTag = Ndef.get(tag);
            NdefMessage ndefMessage = ndefTag.getCachedNdefMessage();
            NdefRecord[] ndefRecords = ndefMessage.getRecords();
            if (ndefRecords.length > 0) {
                for (NdefRecord item : ndefRecords) {
                    try {
                        if (new String(item.getType()).equals(new String(NdefRecord.RTD_TEXT))) {
                            ret.append(new String(item.getPayload(), Charset.forName("UTF-8")).substring("\nko".length()));
                        } else if (new String(item.getType()).equals(new String(NdefRecord.RTD_URI))) {
                            ret.append(new String(item.getPayload(), Charset.forName("UTF-8")));
                        } else /*NdefRecord.RTD_ANDROID_APP*/ if (new String(item.getType()).equals("android.com:pkg")) {
                            ret.append(new String(item.getPayload()));
                        }
                    } catch (Exception ignore) {
                    }
                }
            }
        } catch (Exception ignore) {
        }
        return ret.toString();
    }

    @NotNull
    @Contract("_ -> new")
    public static String bytesToHex(@NotNull byte[] bytes) {
        final int length = bytes.length;
        char[] hexChars = new char[length * 2];
        for (int i = 0; i < length; i++) {
            int var = bytes[i] & 0xFF;
            hexChars[i * 2] = HEX_ARRAY[var >>> 4];
            hexChars[i * 2 + 1] = HEX_ARRAY[var & 0x0F];
        }
        String string = new String(hexChars).replaceAll("(..)", "$1:");
        return string.substring(0, string.length() - 1);
    }

    /**
     * NFC 기능 동작 확인
     * (NFC 꺼짐) 사용자가 NFC 기능을 켤 수 있게 팝업 생성
     */
    public static boolean isNfcEnabled() {
        return nfcAdapter != null && nfcAdapter.isEnabled();
    }

    public void onResume() {
        if (nfcAdapter != null) {
            nfcAdapter.enableForegroundDispatch(activity, pendingIntent, intentFilters, techLists);
        }
    }

    public void onPause() {
        if (nfcAdapter != null) {
            nfcAdapter.disableForegroundDispatch(activity);
        }
    }
}

File: app/src/main/java/kr/djspi/pipe01/nfc/NfcUtil.kt
package kr.djspi.pipe01.nfc

import android.app.Activity
import android.app.PendingIntent
import android.content.Intent
import android.content.IntentFilter
import android.content.IntentFilter.MalformedMimeTypeException
import android.nfc.NdefMessage
import android.nfc.NdefRecord.*
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.nfc.tech.MifareUltralight
import android.nfc.tech.Ndef
import android.nfc.tech.NfcA
import android.util.Log
import com.nxp.nfclib.CardType.NTag213
import com.nxp.nfclib.CardType.NTag216
import com.nxp.nfclib.NxpNfcLib
import com.nxp.nfclib.exceptions.NxpNfcLibException
import com.nxp.nfclib.ndef.NdefMessageWrapper
import com.nxp.nfclib.ndef.NdefRecordWrapper
import com.nxp.nfclib.ntag.INTag213215216
import com.nxp.nfclib.ntag.NTagFactory
import kr.djspi.pipe01.BuildConfig.*
import java.nio.charset.Charset
import java.util.*

class NfcUtil(private val activity: Activity, useActivityClass: Class<*>) {

    private lateinit var intentFilters: Array<IntentFilter>
    private var nxpNfcLib: NxpNfcLib? = null

    /**
     * 아래의 변수들은 반드시 final 선언해야만 하며, 그렇지 않을 경우 intent 들 간의 간섭이 발생하여
     * NFC 태그를 태깅하면 이전 액티비티 인텐트를 실행하기도 한다.
     */
    private val pendingIntent: PendingIntent
    private val techLists: Array<Array<String>>
    private var objNtag: INTag213215216? = null
    private var nfcAdapter: NfcAdapter? = NfcAdapter.getDefaultAdapter(this.activity)

    init {
        try {
            intentFilters = arrayOf(IntentFilter(NfcAdapter.ACTION_TECH_DISCOVERED))
//            intentFilters[0].addDataType("application/kr.djspi.pipe01")
        } catch (ignore: MalformedMimeTypeException) {
        }

        pendingIntent = PendingIntent.getActivity(
            this.activity,
            0,
            Intent(this.activity, useActivityClass).setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP),
            0
        )
        techLists = arrayOf(
            arrayOf(NfcA::class.java.name),
            arrayOf(Ndef::class.java.name),
            arrayOf(MifareUltralight::class.java.name)
        )

        initializeLibrary(this.activity)
    }

    /**
     * TapLinx NTAG 라이브러리를 불러옴
     */
    private fun initializeLibrary(activity: Activity) {
        try {
            nxpNfcLib = NxpNfcLib.getInstance()
            nxpNfcLib?.registerActivity(activity, NFC_APP_KEY, NFC_OFFLINE_KEY)
        } catch (e: Exception) {
        }
    }

    /**
     * NTAG 의 모델명을 찾은 뒤 적절한 태그 객체를 리턴
     * (NTAG 216 이 아님) 에러 메시지 출력, null 리턴
     *
     * @param intent 발생한 NTAG 인텐트
     * @return objNtag 인식된 NTAG 216 태그 객체
     */
    private fun getNtag(intent: Intent): INTag213215216? {
        try {
            objNtag = when {
                nxpNfcLib?.getCardType(intent) == NTag216
                -> NTagFactory.getInstance().getNTAG216(nxpNfcLib?.customModules)

                nxpNfcLib?.getCardType(intent) == NTag213
                -> NTagFactory.getInstance().getNTAG213(nxpNfcLib?.customModules)

                else -> {
                    Log.w("NfcUtil", "Tag is NOT NTAG 216 Type")
                    return null
                }
            }
        } catch (e: NxpNfcLibException) {
            e.message
        }
        return objNtag
    }

    /**
     * 쓰기 대상 태그에 쓰기 작업 수행
     * (setReadOnly) build.gradle 에서 설정: release 버전(true)과 내부용 alpha 버전(false) 구분해 빌드
     *
     * @param intent 전달된 태그 인텐트
     * @return isSuccess 쓰기 작업 성공 여부
     * @see NdefRecordWrapper[] createRecord 사용자 입력값으로 NDEF 레코드를 생성해 리턴
     */
    fun writeTag(intent: Intent, strings: Array<String?>): Boolean {
        var isSuccess = false
        try {
            objNtag = getNtag(intent)
            objNtag?.let {
                it.reader.connect()
                it.reader.timeout = 2000
                it.authenticatePwd(k, a) // 비밀번호 인증
                if (it.isPwdAuthenticated) {
                    // NDEF 메시지 생성 & 쓰기 작업 실행
                    it.writeNDEF(NdefMessageWrapper(createRecord(records = strings)))
                    Log.w("NfcUtil", "NTag Written")
                    if (setReadOnly) it.makeCardReadOnly() // 쓰기 후 Read-Only 로 설정
                    isSuccess = true
                }
            }
        } catch (e: NullPointerException) {
            Log.e("NfcUtil", "Exception Thrown: writeTag()")
            e.message
        } catch (e: NxpNfcLibException) {
            Log.e("NfcUtil", "Exception Thrown: writeTag()")
            e.message
        } catch (e: IllegalArgumentException) {
            Log.e("NfcUtil", "Exception Thrown: writeTag()")
            e.message
        }
        objNtag?.reader?.close()
        return isSuccess
    }

    /**
     * (TapLinx 라이브러리) NDEF 레코드 생성함수
     * 0번 레코드(고정): 앱 패키지명 지정. SPI 정품 태그를 태깅할 시
     * 앱이 설치되어 있으면 앱 실행, 앱이 설치되어 있지 않으면 플레이스토어 설치페이지로 연결됨
     * 1번 레코드~: SPI 정보
     * 마지막 레코드(고정): 태깅된 NFC 칩의 UID(시리얼번호) 지정
     *
     * @param locale 텍스트 데이터 로컬 설정: (ko)
     * @param utf8 UTF-8 인코딩으로 기록
     * @return wrappers 태그에 기록되는 레코드 Array
     */
    private fun createRecord(
        locale: Locale = Locale.KOREAN,
        utf8: Boolean = true,
        records: Array<String?>
    ): Array<NdefRecordWrapper?> {
        val encoding: Charset = if (utf8) Charsets.UTF_8 else Charsets.UTF_16
        val length: Int = records.size
        val texts: Array<ByteArray?> = arrayOfNulls(length)
        val data: Array<ByteArray?> = arrayOfNulls(length)
        val langBytes = locale.language.toByteArray(Charsets.UTF_8)
        val langBytesLength: Int = langBytes.size
        val status: Char = (0 + langBytesLength).toChar()
        val wrappers = arrayOfNulls<NdefRecordWrapper>(length + 1)
        for (i in 0 until length) {
            texts[i] = records[i]!!.toByteArray(encoding)
            data[i] = ByteArray(langBytesLength + texts[i]!!.size + 1) // 0번 레코드를 위해 1자리 추가
            data[i]!![0] = status.code.toByte()
            System.arraycopy(langBytes, 0, data[i]!!, 1, langBytesLength)
            System.arraycopy(texts[i] as Any, 0, data[i]!!, 1 + langBytesLength, texts[i]!!.size)
            wrappers[i + 1] = NdefRecordWrapper(
                TNF_WELL_KNOWN,
                RTD_TEXT,
                i.toString().toByteArray(),
                data[i]
            )
        }
        // 1번 레코드부터 생성하는 for 문이 끝난 뒤 0번 레코드(패키지명)를 생성한다.
        wrappers[0] = NdefRecordWrapper(createApplicationRecord(APPLICATION_ID))
        return wrappers
    }

    /**
     * 태그에 기록된 정보를 Record 단위로 불러오며, Record index 와 ArrayList 의 index 는 1:1 대응된다.
     * (정보가 기록되지 않음) null 을 리턴
     *
     * @return 기록된 정보를 담은 ArrayList<String>
     * @throws NullPointerException (태그 없음) 정보를 가져오는 과정에서 null 참조
    </String> */
    @Throws(NullPointerException::class)
    fun getRecord(intent: Intent): ArrayList<String> {
        val recordList = ArrayList<String>()
        val rawMsg = intent.getParcelableArrayExtra(NfcAdapter.EXTRA_NDEF_MESSAGES)
        rawMsg?.let {
            val records = NdefMessageParser.parse(it[0] as NdefMessage)
            recordList.add(0, "")
            val listLength = records.size + 1
            for (i in 1 until listLength) {
                recordList.add(i, (records[i - 1] as TextRecord).text)
            }
        }
        return recordList
    }

    @Throws(NullPointerException::class)
    fun getSerial(intent: Intent): String {
        val tag: Tag = intent.getParcelableExtra(NfcAdapter.EXTRA_TAG)
            ?: throw NullPointerException("Not Found")
        val bytes: ByteArray = tag.id
        val stringBuilder = StringBuilder(0)
        for (byte in bytes) {
            stringBuilder.append(String.format("%02X", byte))
        }
        val serial = stringBuilder.toString().replace("(..)".toRegex(), "$1:")
        return serial.substring(0, serial.length - 1)
    }

    /**
     * NFC 기능 동작 확인
     * (NFC 꺼짐) 사용자가 NFC 기능을 켤 수 있게 팝업 생성
     */
    fun isNfcEnabled(): Boolean {
        return nfcAdapter != null && nfcAdapter!!.isEnabled
    }

    fun onNewTagIntent(intent: Intent): Tag {
        return intent.getParcelableExtra(NfcAdapter.EXTRA_TAG)!!
    }

    fun onResume() {
        nfcAdapter?.enableForegroundDispatch(activity, pendingIntent, intentFilters, techLists)
    }

    fun onPause() {
        nfcAdapter?.disableForegroundDispatch(activity)
    }

    /**
     * Reserved for spi-init
     *
    @Throws(IOException::class, FormatException::class)
    fun writeInitTag(
    intent: Intent,
    strings: Array<String?>
    ): Boolean {
    val length = strings.size
    val ndefRecords = arrayOfNulls<NdefRecord>(length)
    ndefRecords[0] = createApplicationRecord(strings[0])
    for (i in 1 until length) {
    ndefRecords[ i ] = createTextRecord(null, strings[ i ])
    }
    val ndefMessage = NdefMessage(ndefRecords)

    // 초기화 정보를 기록
    val tag = intent.getParcelableExtra<Tag>(NfcAdapter.EXTRA_TAG)
    val ndef = Ndef.get(tag)
    if (ndef != null) {
    ndef.connect()
    if (ndef.isWritable && ndef.maxSize >= ndefMessage.toByteArray().size) {
    ndef.writeNdefMessage(ndefMessage)
    ndef.close()
    return true
    }
    } else {
    val format = NdefFormatable.get(tag)
    if (format != null) {
    format.connect()
    format.format(ndefMessage)
    format.close()
    return true
    }
    }
    return false
    }

    fun readTag(tag: Tag): String {
    val ret = StringBuilder()
    try {
    val ndefTag = Ndef.get(tag)
    val ndefMessage = ndefTag.cachedNdefMessage
    val ndefRecords = ndefMessage.records
    if (ndefRecords.isNotEmpty()) {
    ndefRecords.forEach {
    when {
    String(it.type) == String(RTD_TEXT)
    -> ret.append(String(it.payload, Charsets.UTF_8).substring("\nko".length))
    String(it.type) == String(RTD_URI)
    -> ret.append(String(it.payload, Charsets.UTF_8))
    String(it.type) == "android.com:pkg"
    -> ret.append(String(it.payload))
    }
    }
    }
    } catch (ignore: Exception) {
    }
    return ret.toString()
    }

    fun setNtagPwdProtection(intent: Intent, setReadOnly: Boolean): Boolean {
    return try {
    val tag: INTag213215216? = getNtag(intent)
    tag?.programPWDPack(k, a)
    tag?.enablePasswordProtection(false, 4)
    if (setReadOnly) {
    tag?.makeCardReadOnly()
    }
    true
    } catch (e: Exception) {
    false
    }
    }

    @Throws(NullPointerException::class, NxpNfcLibException::class)
    fun clearTag(tag: INTag213215216?): Boolean {
    if (tag == null) {
    throw NullPointerException("Not Found")
    }
    tag.authenticatePwd(k, a)
    return if (tag.isPwdAuthenticated) {
    tag.enablePasswordProtection(false, 255)
    tag.clear()
    true
    } else {
    throw NxpNfcLibException("Auth Fail")
    }
    }

    @Throws(NullPointerException::class, NxpNfcLibException::class)
    fun getCardType(intent: Intent): CardType? {
    val type = nxpNfcLib?.getCardType(intent)
    if (type == CardType.UnknownCard) {
    throw NxpNfcLibException("Bad Type")
    }
    return when (type) {
    NTag213 -> NTag213
    NTag216 -> NTag216
    else -> throw IllegalArgumentException("Bad Type")
    }
    }
     *
     */
}
